smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java

```diff
@@ -250,6 +254,47 @@ public Jid getArchiveAddress() {
         return archiveAddress;
     }
 
+    /**
+     * Returns the MAM namespace used by this {@link MamManager}. If the archive does not support any MAM namespace
+     * supported by Smack, null is returned.
+     *
+     * @return the MAM namespace used by this manager, null if MAM is not supported
+     * @throws NoResponseException if there was no response from the remote entity.
+     * @throws XMPPErrorException if there was an XMPP error returned.
+     * @throws NotConnectedException if the XMPP connection is not connected.
+     * @throws InterruptedException if the calling thread was interrupted.
+     */
+    public String getMamNamespace() throws XMPPErrorException, NotConnectedException, NoResponseException, InterruptedException {
+        MamVersion mamVersion = getSupportedMamVersionOrNull();
+        return mamVersion == null ? null : mamVersion.getNamespace();
+    }
+
+    private MamVersion getSupportedMamVersionOrNull() throws XMPPErrorException, NotConnectedException, NoResponseException, InterruptedException {
+        if (mamVersion == null) {
+            DiscoverInfo info = serviceDiscoveryManager.discoverInfo(getArchiveAddress());
+            for (MamVersion v : MamVersion.values()) {
```

Comment:
Shouldn't we somehow ensure that we always detect the best, which usually means latest, version in case multiple are supported. I am not sure about the contract of an Enum's `values()`, but I think it could be implementation dependent. If so, then this would currently potentially yield different results in different Java runtime environments.

Reply:
According to the Java Language Specification values() returns in the order the values were declared. So I think we can safely assume the order to be as we expect. See: https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.9.3.

Reply:
I did add some code comments to clarify this.